bip-xxxx.mediawiki

```diff
@@ -0,0 +1,303 @@
+|out-of-utxos
+|The receiver does not have any UTXO to contribute in a payjoin proposal.
+|-
+|not-enough-money
+|The receiver added some inputs but could not bump the fee of the payjoin proposal.
+|-
+|insane-psbt
+|Some consistency check on the PSBT failed.
+|-
+|version-unsupported
+|This version of payjoin is not supported.
+|-
+|need-utxo-information
+|The witness UTXO or non witness UTXO is missing
+|}
+
+The receiver is allowed to return implementation specific errors which may assist the sender to diagnose any issue.
+
+However, it is important that error codes that are not well-known and that the message do not appear on the sender's software user interface.
+Such error code or message could be used maliciously to fish a non technical user.
+Instead those error or message can only appear in debug logs.
+
+It is advise to hard code the description of the error code into the sender's software.
+
+===Sender's payjoin proposal checklist===
+
+The sender should check the payjoin proposal before signing it to prevent a malicious receiver from stealing money.
+
+* Check that all the spent outpoints in the original PSBT still exists in the coinjoin PSBT.
+* Check that all the spent outpoints in the original PSBT does not have any partial signature.
+* If the sender is not using inputs with mixed types, check that the receiver inputs type match the inputs type of the sender. (ie. both using P2SH-P2WPKH or both using P2WPKH)
+* Check that any inputs added by the receiver are finalized.
+* Check that the transaction version, and nLockTime are unchanged.
+* Check that the sender's inputs' sequence numbers are unchanged.
+* Check that the sent amount in the payjoin proposal is less or equal to the sent amount of the original transaction.
+
+If the sent amount in the payjoin proposal is above the amount sent in the original PSBT
+* Check that the additional paid amount has been paid to fee.
+* Check that the additional paid amount is not more than twice the original fee 
+* Check that the additional paid amount has only been used to pay the receiver's additional inputs.
+
+The sender must be careful to only sign the inputs that were present in the original PSBT and nothing else.
+
+Note:
+
+* The sender **does NOT check** whether ouputs have been removed or modified. This allows flexibility to the receiver to adapt his receiving address type to match the other outputs's address type of the sender, or, on the contrary, to create a payment output which would be considered a change address by common chain analysis heuristic. For example, if the receiver supports both P2WPKH and P2SH-P2WPKH, even if the invoice's address in the original transaction was P2WPKH, the receiver may change the address to be P2SH-P2WPKH to match sender's change address format. This is safe because the sender only cares that they do not send too much money in the payjoin transaction. It is also useful if the receiver wants to batch some of their own payments in the transaction.
```

Comment:
> What if the receiver replaces the change outputs' scriptPubKey by another already known address of the sender, forcing the sender to reuse addresses and damaging his privacy? Or Give it back some dust change even when the sender decided to spend the whole coin (to the same utxo scriptPubKey) or things like that.

We should tighten this. Indeed, the sender may want to do a batch payment for example.
However it is important that the receiver can change his payment output. For example, if the receiver has 2 wallets and the payer is using P2SH for paying a P2WPKH, then the receiver should be able to switch the payment output to P2SH for improved privacy.

> Receiver could decide to split the change output in some way that only makes the sender to pay more fee.

I enumerated the checks that the sender need to do to make sure he does not pay too much fee. (It is capped to twice the fee)
Moreover, it is important for the receiver to also be able to batch his own payments.

Reply:
> What if the receiver replaces the change outputs' scriptPubKey by another already known address of the sender, forcing the sender to reuse addresses and damaging his privacy? Or Give it back some dust change even when the sender decided to spend the whole coin (to the same utxo scriptPubKey) or things like that.

Is there perhaps a misunderstanding here @lontivero ? I took this ``does NOT check`` section to mean only that the sender does not check the type or number of the receiver's outputs - the sender certainly *does* check his own output, insisting on an unchanged scriptPubKey, and of course checking that the output amounts follow the other rules described.

But, the language should be clearer. It is the receiver's outputs that don't need to be checked.

Reply:
Is https://github.com/NicolasDorier/bips/pull/3/commits/50bda8ee22e11f1be8e0b18450ab8a394ca4662e better?

Reply:
Probably I misunderstood this @AdamISZ, what you say makes sense to me.
@NicolasDorier 50bda8e is more clear. I am not sure my point is really important because the sender can always reject a proposal from the receiver.